libs/widget-lib/src/messages.ts

```diff
@@ -35,25 +35,50 @@ export function postMessageToWindow(contentWindow: Window, method: string, paylo
   )
 }
 
+export type WindowListener = (event: MessageEvent<unknown>) => void
+
 export function listenToMessageFromWindow<T extends WidgetMethodsListen>(
   contentWindow: Window,
   method: T,
   callback: (payload: WidgetMethodsListenPayloadMap[T]) => void
-): void
+): WindowListener
 export function listenToMessageFromWindow<T extends WidgetMethodsEmit>(
   contentWindow: Window,
   method: T,
   callback: (payload: WidgetMethodsEmitPayloadMap[T]) => void
-): void
+): WindowListener
 
-export function listenToMessageFromWindow(contentWindow: Window, method: string, callback: (payload: unknown) => void) {
-  contentWindow.addEventListener('message', (event) => {
-    if (event.data.key !== COW_SWAP_WIDGET_EVENT_KEY || event.data.method !== method) {
+export function listenToMessageFromWindow<T = unknown>(
+  contentWindow: Window,
+  method: string,
+  callback: (payload: T) => void
+): (payload: MessageEvent<unknown>) => void {
+  const listener = (event: MessageEvent<unknown>) => {
+    if (!isEventData(event.data) || event.data.key !== COW_SWAP_WIDGET_EVENT_KEY || event.data.method !== method) {
       return
     }
 
-    callback(event.data)
-  })
+    callback(event.data as T)
+  }
+  contentWindow.addEventListener('message', listener)
+
+  return listener
+}
+
+interface EventData {
+  key: string
+  method: string
+}
+
+function isEventData(obj: unknown): obj is EventData {
+  return (
+    typeof obj === 'object' &&
+    obj !== null &&
```

Comment:
Nitpick: move the check up and broaden it to any `falsy` value:

```suggestion
    !!obj &&
    typeof obj === 'object' &&
```

I guess that works?

Reply:
Not sure if you refer to doing this, but we need to assert is an object: 

<img width="948" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/a8a25030-74a4-41c0-b4b1-adf604f19bb2">

Reply:
No, I just suggested to move the null check up.
Pretty silly sorting suggestion, you can ignore it.